drivers/xen/gnttab.c

```diff
@@ -0,0 +1,220 @@
+    BUG_ON(ref < NR_RESERVED_ENTRIES || ref >= NR_GRANT_ENTRIES);
+    gnttab_list[0] = gnttab_list[ref];
+    local_irq_restore(flags);
+    return ref;
+}
+
+grant_ref_t gnttab_grant_access(domid_t domid, unsigned long frame, int readonly)
+{
+    grant_ref_t ref;
+
+    ref = get_free_entry();
+    gnttab_table[ref].frame = frame;
+    gnttab_table[ref].domid = domid;
+    wmb();
+    readonly *= GTF_readonly;
+    gnttab_table[ref].flags = GTF_permit_access | readonly;
+
+    return ref;
+}
+
+grant_ref_t gnttab_grant_transfer(domid_t domid, unsigned long pfn)
+{
+    grant_ref_t ref;
+
+    ref = get_free_entry();
+    gnttab_table[ref].frame = pfn;
+    gnttab_table[ref].domid = domid;
+    wmb();
+    gnttab_table[ref].flags = GTF_accept_transfer;
+
+    return ref;
+}
+
+int gnttab_end_access(grant_ref_t ref)
+{
+    uint16_t flags, nflags;
+
+    BUG_ON(ref >= NR_GRANT_ENTRIES || ref < NR_RESERVED_ENTRIES);
+
+    nflags = gnttab_table[ref].flags;
+    do {
+        if ((flags = nflags) & (GTF_reading|GTF_writing)) {
+            printk("WARNING: g.e. still in use! (%x)\n", flags);
+            return 0;
+        }
+    } while ((nflags = synch_cmpxchg(&gnttab_table[ref].flags, flags, 0)) !=
+            flags);
+
+    put_free_entry(ref);
+    return 1;
+}
+
+unsigned long gnttab_end_transfer(grant_ref_t ref)
+{
+    unsigned long frame;
+    uint16_t flags;
+
+    BUG_ON(ref >= NR_GRANT_ENTRIES || ref < NR_RESERVED_ENTRIES);
+
+    while (!((flags = gnttab_table[ref].flags) & GTF_transfer_committed)) {
+        if (synch_cmpxchg(&gnttab_table[ref].flags, flags, 0) == flags) {
+            printk("Release unused transfer grant.\n");
+            put_free_entry(ref);
+            return 0;
+        }
+    }
+
+    /* If a transfer is in progress then wait until it is completed. */
+    while (!(flags & GTF_transfer_completed)) {
+        flags = gnttab_table[ref].flags;
+    }
+
+    /* Read the frame number /after/ reading completion status. */
+    rmb();
+    frame = gnttab_table[ref].frame;
+
+    put_free_entry(ref);
+
+    return frame;
+}
+
+grant_ref_t gnttab_alloc_and_grant(void **map)
+{
+    unsigned long mfn;
+    grant_ref_t gref;
+
+    *map = (void *)memalign(PAGE_SIZE, PAGE_SIZE);
+    mfn = virt_to_mfn(*map);
+    gref = gnttab_grant_access(0, mfn, 0);
+    return gref;
+}
+
+static const char * const gnttabop_error_msgs[] = GNTTABOP_error_msgs;
+
+const char *gnttabop_error(int16_t status)
+{
+    status = -status;
+    if (status < 0 || status >= ARRAY_SIZE(gnttabop_error_msgs))
+        return "bad status";
+    else
+        return gnttabop_error_msgs[status];
+}
+
+void init_gnttab(void)
+{
+    struct xen_add_to_physmap xatp;
+    struct gnttab_setup_table setup;
+    xen_pfn_t frames[NR_GRANT_FRAMES];
+    int i, rc;
+
+    debug("%s\n", __func__);
+
+    for (i = NR_RESERVED_ENTRIES; i < NR_GRANT_ENTRIES; i++)
+        put_free_entry(i);
+
+    gnttab_table = memalign(PAGE_SIZE, NR_GRANT_FRAMES * PAGE_SIZE);
```

Comment:
AFAIU, this will allocate part of the RAM for the grant-table. Will this region be marked as unusable for the DomU OS afterwards?

Reply:
When jumping to the kernel we remove all the mappings, so the kernel will setup its own gnttab, so no need to mark any region

Reply:
Would you mind to give me a pointer to that code removing the grant table frames?

Reply:
https://github.com/xen-troops/u-boot/pull/1/commits/f065d0371da9e15e19fd91b9d71604e00814ff6c#diff-38af42740ed5d67089e2992db6a1a39cR209
Please note that `fini_gnttab` is not called in this pull request, but it is in the new one

Reply:
Well, the hypercall **GNTTABOP_setup_table** doesn't behave as you think. For Arm, it will only ensure there is enough frames allocated in the hypervisor.  There is no management of the mappings in the p2m.

This is the guest job to map/unamp them using the appropriate hypercalls. The grant-table frames as the vCPU shared pages are memory from Xen mapped into the guest address space. So the proper way for map/unmap is the following:

-  For the mapping 
1. Remove the RAM mapping using **XENMEM_decrease_reservation**
2. Map the grant frame using **XENMEM_add_to_physmap**

- For the unmapping
1. Unmap the grant frame using **XENMEM_remove_from_physmap**
2. Populate the region using **XENMEM_populate_physmap**

Step 1 of the mapping is required to avoid losing the RAM page for the lifetime of your guest.
Step 2 of the unmapping is required so the address point again to a RAM page.

Reply:
Probably I miss your point, but we do not map and unmap any foreign grant references, we only share our own pages with the backend.
For the shared page we use:
```
	xatp.space = XENMAPSPACE_shared_info;
	xatp.gpfn = virt_to_pfn(HYPERVISOR_shared_info);
	if (HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp) != 0)

```
For the grant-table we use:
```
        xatp.space = XENMAPSPACE_grant_table;
        xatp.gpfn = ((unsigned long)gnttab_table >> PAGE_SHIFT) + i;
        rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
...
    setup.nr_frames = NR_GRANT_FRAMES;
    set_xen_guest_handle(setup.frame_list, frames);
    rc = HYPERVISOR_grant_table_op(GNTTABOP_setup_table, &setup, 1);

```
When we setup a shared ring we:
```
	SHARED_RING_INIT(s);
...
	dev->ring_ref = gnttab_grant_access(dev->dom, virt_to_pfn(s), 0);

```
While pvblock shares the pages we use:
```
			gnttab_grant_access(...),
			gnttab_end_access(...);

```
So, as we do not have any foreign mappings, why would we need anything else like you mentioned?

Reply:
> Probably I miss your point, but we do not map and unmap any foreign grant references, we only share our own pages with the backend.

Notice that I mentioned memory shared with **xen** and not the **backend**. The way the two works are fairly different see more below.

>
> For the shared page we use:
> 
> ```
> 	xatp.space = XENMAPSPACE_shared_info;
> 	xatp.gpfn = virt_to_pfn(HYPERVISOR_shared_info);
> 	if (HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp) != 0)
> ```

For this one and ...

> 
> For the grant-table we use:
> 
> ```
>         xatp.space = XENMAPSPACE_grant_table;
>         xatp.gpfn = ((unsigned long)gnttab_table >> PAGE_SHIFT) + i;
>         rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
> ```

... this one, they are used to map **xen** memory into the guest. IOW, the P2M entry associated to the field *xatp.gfn* will be updated to point to **xen** memory. If there was any mapping (such as to a RAM page), then it will disappear forever. Effectively, you will leave a hole in your address space.

But I would strongly recommend to avoid using RAM page for the grant-table if you can. This will end up to shatter superpages in your P2M and potentially affect the performance of your guest long-term. You will be better off to use the region provided by the device-tree.

Reply:
Ok, this can be done (grant table from the device tree). And if we use those then we don't need to play with reservations, right?

Reply:
You will not need to remove the previous mapping and repopulate it, but you will still have to unmap whatever you map to stay on the _safe_ side.

Reply:
Thank you, we have implemented the gnttab with the use of device tree reg and XENMEM_add_to_physmap/XENMEM_remove_from_physmap and without XENMEM_decrease_reservation/XENMEM_populate_physmap.
Much cleaner now.